Introduce new syntax for overriding selector (#968)
hyperledger-labs/solang: docs/language/functions.rst

    for storage rent and there is a minimum deposit needed for the contract. As a result,
    constructors always receive value on Parity Substrate.

Overriding function selector
____________________________

> When a function is called, the function selector and the arguments to the function

Reviewer: ```suggestion
When a function is called, its selector and its arguments
```
Function was repeated three times here.

Author: I don't think "its selector" is particularly clear. It's a called a function selector.

Author: Also now you repeated its

Author: I've rephrased this, please let me know what you think.

are serialized (also known as abi encoded) and passed to the program. The selector
is what the runtime program uses to determine what function was called. Usually
the function selector is generated from a deterministic hash value of the function
name and its arguments.

The selector value can be overriden with the ``selector=hex"deadbea1"`` syntax,
for example:

.. code-block:: solidity

    contract foo {

>         function get_foo() selector=hex"01020304" public return (int) {

Reviewer 2: Does the position of the selector overwrite matter? I'd rather move the selector overwrite at the beginning or at the end of the function declaration.

Author: The selector is part of the function attributes, which can occur in any order after the parameters and before the `returns (..)` (if any).

            return 102;
        }
    }

Only ``public`` and ``external`` functions have a selector, and can have their
selector overriden. On Substrate, constructors have selectors too, so they
can also have their selector overriden. If a function overrides another one in a
base contract, then the selector of both must match.

.. warning::
    On ewasm and Solana, changing the selector may result in a mismatch between

>     the contract metadata and the actual contract code, because the former does not explicitly store the selector, but derives it

Reviewer 2: Is the issue here when you derive the metadata out of a contract already deployed, resulting in wrong metadata? If that is possible and common for Solana, then overwriting sounds super scary to me...

Author: This currently very messy on Solana. Currently there are many contracts without metadata which have hand-coded selectors (the idea of metadata is fairly new for Solana!). This is why we need the selector override in the first place. They need hand-written solidity `interface { ... }` definitions with the selector override.

We're adding selectors to the solana metadata to solve all these problems, and this scary warning will go away for Solana too.

Reviewer 2: Ah I see :grin: I mean selector overwrites could maybe used for advanced things like hooks. And it is possible in `ink!` anyways. So I think in general it is still a useful feature to have.

    not explicitly store the selector.

>     Only use this feature if you know what you are doing.

Reviewer: Instead of saying `Only use this feature if you know what you are doing`, I think it is more polite to say `Use this feature carefully, as it may either break a contract or cause undefined behavior.` What do you think?

Author: I've added this, thanks

    for storage rent and there is a minimum deposit needed for the contract. As a result,
    constructors always receive value on Parity Substrate.

Overriding function selector
____________________________

When a function is called, the function selector and the arguments to the function
are serialized (also known as abi encoded) and passed to the program. The selector
is what the runtime program uses to determine what function was called. Usually
the function selector is generated from a deterministic hash value of the function
name and its arguments.

To change the selector value, it can be overriden with the ``selector=hex"deadbea1"``
syntax, for example:

.. code-block:: solidity

    contract foo {
        function get_foo() selector=hex"01020304" public return (int) {
            return 102;
        }
    }

Only ``public`` and ``external`` functions have a selector, and can have their
selector overriden. On Substrate, constructors have selectors too, so they have
can also have their selector overriden. If a function overrides a function in a

> base contract, then the selector of the function in the base contract must match.

Author: Good point, thanks

.. warning::

>     On ewasm and Solana, changing the selector may result in a mismatch between
>     the contract metadata and the actual contract code. This is because the ewasm
>     and Solana metadata does not explicitly store the selector, but derives it

Reviewer: ```suggestion
    On ewasm and Solana, changing the selector may result in a mismatch between
    the contract metadata and the actual contract code, because the former does not explicitly store the selector, but derives it
```

Reviewer: A suggestion to make it less wordy.
